OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs

```diff
@@ -0,0 +1,234 @@
+			"Only applies if RestrictMCVDeploymentFallbackToBase is enabled and there's at least one construction yard.")]
+		public readonly int MaxBaseRadius = 20;
+
+		[Desc("Should deployment of additional MCVs be restricted to MaxBaseRadius if explicit deploy locations are missing or occupied?")]
+		public readonly bool RestrictMCVDeploymentFallbackToBase = true;
+
+		public override object Create(ActorInitializer init) { return new McvManagerBotModule(init.Self, this); }
+	}
+
+	public class McvManagerBotModule : ConditionalTrait<McvManagerBotModuleInfo>, IBotTick, IBotPositionsUpdated
+	{
+		public CPos GetRandomBaseCenter()
+		{
+			var randomConstructionYard = world.Actors.Where(a => a.Owner == player &&
+				Info.ConstructionYardTypes.Contains(a.Info.Name))
+				.RandomOrDefault(world.LocalRandom);
+
+			return randomConstructionYard != null ? randomConstructionYard.Location : initialBaseCenter;
+		}
+
+		readonly World world;
+		readonly Player player;
+
+		readonly Predicate<Actor> unitCannotBeOrdered;
+
+		IBotPositionsUpdated[] notifyPositionsUpdated;
+		IBotRequestUnitProduction[] requestUnitProduction;
+
+		CPos initialBaseCenter;
+		int scanInterval;
+		int ticks;
+
+		// MCVs that the bot already knows about. Any MCV not on this list needs to be given an order.
+		List<Actor> activeMCVs = new List<Actor>();
+
+		public McvManagerBotModule(Actor self, McvManagerBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+
+			if (world.Type == WorldType.Editor)
+				return;
+
+			unitCannotBeOrdered = a => a.Owner != player || a.IsDead || !a.IsInWorld;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			notifyPositionsUpdated = player.PlayerActor.TraitsImplementing<IBotPositionsUpdated>().ToArray();
+			requestUnitProduction = player.PlayerActor.TraitsImplementing<IBotRequestUnitProduction>().ToArray();
+
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			scanInterval = world.LocalRandom.Next(Info.ScanForNewMcvInterval, Info.ScanForNewMcvInterval * 2);
+		}
+
+		void IBotPositionsUpdated.UpdatedBaseCenter(CPos newLocation)
+		{
+			initialBaseCenter = newLocation;
+		}
+
+		void IBotPositionsUpdated.UpdatedDefenseCenter(CPos newLocation) { }
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			ticks++;
+
+			if (ticks == 1)
+				DeployMcvs(bot, false);
+
+			if (--scanInterval <= 0)
+			{
+				scanInterval = Info.ScanForNewMcvInterval;
+				DeployMcvs(bot, true);
+
+				// No construction yards - Build a new MCV
+				if (ShouldBuildMCV)
+				{
+					var unitBuilder = requestUnitProduction.FirstOrDefault(Exts.IsTraitEnabled);
+					if (unitBuilder != null)
+					{
+						var mcvInfo = AIUtils.GetInfoByCommonName(Info.McvTypes, player);
+						unitBuilder.RequestUnitProduction(bot, mcvInfo.Name);
+					}
+				}
+			}
+		}
+
+		bool ShouldBuildMCV
```

Comment:
Please make this a method instead of a property.  Generally speaking, get-properties are used in cases where the result can be thought of like a field: so no side-effects, and low performance overheads.  `CountActorByCommonName` can be quite heavy, so this breaks the second expectation.